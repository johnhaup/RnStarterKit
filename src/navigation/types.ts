import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

export enum Tabs {
  TAB_ONE = 'TAB_ONE',
  TAB_TWO = 'TAB_TWO',
}

export enum Modals {
  A_MODAL = 'A_MODAL',
}

export enum Screens {
  // TabOne
  A_SCREEN = 'A_SCREEN',
  B_SCREEN = 'B_SCREEN',
  // TabTwo
  C_SCREEN = 'C_SCREEN',
  D_SCREEN = 'D_SCREEN',
}

export type ScreenParams = {
  [Modals.A_MODAL]: { modalNavProp: string };
  [Screens.A_SCREEN]: undefined;
  [Screens.B_SCREEN]: { navProp?: string } | undefined;
  [Screens.C_SCREEN]: undefined;
  [Screens.D_SCREEN]: { navProp?: string } | undefined;
};

export type NavigationProps<T extends keyof ScreenParams> = {
  navigation: StackNavigationProp<ScreenParams, T>;
  route: RouteProp<ScreenParams, T>;
};
