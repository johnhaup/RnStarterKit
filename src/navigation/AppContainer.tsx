import {
  NavigationContainer,
  NavigationState,
  PartialState,
} from '@react-navigation/native';
import React, { useRef } from 'react';
import { SplashScreen } from '../screens/loading/SplashScreen';
import { useTypedStoreState } from '../store/hooks';
import { RootNavigator } from './navigators/RootNavigator';

// Gets the current screen from navigation state
const getActiveRouteName = (
  state: NavigationState | PartialState<NavigationState>,
): string => {
  const route = state.index ? state.routes[state.index] : null;

  if (route?.state) {
    // Dive into nested navigators
    return getActiveRouteName(route.state);
  }

  return route?.name || 'No route name';
};

export const AppContainer = () => {
  const routeNameRef = useRef<string>('');
  const loadingComplete = useTypedStoreState((s) => s.session.loadingComplete);

  const onStateChange = (state: NavigationState | undefined) => {
    if (!state) {
      return;
    }
    const previousRouteName = routeNameRef.current;
    const currentRouteName = getActiveRouteName(state);

    if (previousRouteName !== currentRouteName) {
      // Nav state can change without changing a route
      // Only log when screen changes
      // DO SOMETHING WITH CURRENT ROUTE NAME
    }

    routeNameRef.current = currentRouteName;
  };

  return (
    <NavigationContainer onStateChange={onStateChange}>
      {loadingComplete ? <RootNavigator /> : <SplashScreen />}
    </NavigationContainer>
  );
};
