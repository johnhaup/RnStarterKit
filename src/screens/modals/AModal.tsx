import React from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';
import { Modals, NavigationProps } from '../../navigation/types';
import { measurements } from '../../styles';

export const AModal = ({
  navigation,
  route,
}: NavigationProps<Modals.A_MODAL>) => {
  const modalNavProp = route.params?.modalNavProp;

  const onPress = () => navigation.goBack();

  return (
    <View style={styles.container}>
      <View style={styles.modal}>
        <Text>A Modal</Text>
        <Text>{modalNavProp}</Text>
        <Button onPress={onPress} title={'Dismiss'} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  modal: {
    backgroundColor: 'white',
    justifyContent: 'center',
    alignItems: 'center',
    width: measurements.SCREEN_WIDTH - 48,
    height: measurements.SCREEN_HEIGHT * 0.5,
    borderRadius: 8,
  },
});
