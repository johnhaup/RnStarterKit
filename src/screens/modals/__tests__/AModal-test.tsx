/**
 * @format
 */

import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import React from 'react';
import 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { Modals, ScreenParams } from '../../../navigation/types';
import { AModal } from '../AModal';

const navigation = {} as StackNavigationProp<ScreenParams, Modals>;
const route = {} as RouteProp<ScreenParams, Modals>;

it('renders correctly', () => {
  const snap = renderer.create(
    <AModal navigation={navigation} route={route} />,
  );
  expect(snap).toMatchSnapshot();
});
