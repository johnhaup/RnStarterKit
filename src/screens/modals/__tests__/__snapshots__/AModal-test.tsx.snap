// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 8,
        "height": 667,
        "justifyContent": "center",
        "width": 702,
      }
    }
  >
    <Text>
      A Modal
    </Text>
    <Text />
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Dismiss
        </Text>
      </View>
    </View>
  </View>
</View>
`;
